core/arch/aarch64/codec.c

```diff
@@ -2580,6 +2580,225 @@ decode_common(dcontext_t *dcontext, byte *pc, byte *orig_pc, instr_t *instr)
+            ldstex_end = i + 1;
+    }
+    if (i < N) {
+        instr_reset(dcontext, &ibuf[i]);
+        len = i;
+    } else
+        len = N;
+
+    /* Quick check for hopeless situations. */
+    if (len == 0 || !(seen_ldex && seen_stex) ||
+        !(seen_branch_to_start || (instr_is_exclusive_load(&ibuf[0]) ||
+                                   instr_is_exclusive_store(&ibuf[0])))) {
+        for (i = 0; i < len; i++)
+            instr_reset(dcontext, &ibuf[i]);
+        return NULL;
+    }
+
+    /* There are several ways we could choose a sub-block containing both ldex
+     * and stex from the first N instructions. Investigate further, perhaps.
+     * We have already set ldstex_beg and ldstex_end.
+     */
+    ASSERT(ldstex_beg != -1 && ldstex_end != -1 && ldstex_beg < ldstex_end);
+
+    /* Expand ldstex sub-block until it is a single-entry single-exit block. */
+    for (;;) {
+        int new_beg = ldstex_beg;
+        int new_end = ldstex_end;
+        for (i = ldstex_beg; i < ldstex_end; i++) {
+            instr_t *instr = &ibuf[i];
+            if (instr_is_ubr_arch(instr) || instr_is_cbr_arch(instr)) {
+                int target = (uint *)instr_get_branch_target_pc(instr) - pc;
+                if (target > len) {
+                    failed = true;
+                    break;
+                }
+                if (target < new_beg)
+                    new_beg = target;
+                if (target > new_end)
+                    new_end = target;
+            }
+        }
+        if (new_beg == ldstex_beg && new_end == ldstex_end)
+            break;
+        ldstex_beg = new_beg;
+        ldstex_end = new_end;
+    }
+
+    if (ldstex_beg != 0)
+        failed = true;
+
+    if (!failed) {
+        /* Check whether the sub-block uses the stolen register and all of X0-X5.
+         * If it does, it would be impossible to mangle it so it is better not to
+         * create an OP_ldstex.
+         */
+        reg_id_t regs[] = { dr_reg_stolen,
+                            DR_REG_X0, DR_REG_X1, DR_REG_X2,
+                            DR_REG_X3, DR_REG_X4, DR_REG_X5 };
+        int r;
+        for (r = 0; r < sizeof(regs) / sizeof(*regs); r++) {
+            for (i = ldstex_beg; i < ldstex_end; i++) {
+                if (instr_uses_reg(&ibuf[i], regs[r]))
+                    break;
+            }
+            if (i >= ldstex_end)
+                break;
+        }
+        if (r >= sizeof(regs) / sizeof(*regs))
+            failed = true;
+    }
+
+    if (!failed) {
+        instr_create_ldstex(dcontext, ldstex_end - ldstex_beg,
+                            pc + ldstex_beg, &ibuf[ldstex_beg], instr_ldstex);
+    }
+
+    for (i = 0; i < len; i++)
+        instr_reset(dcontext, &ibuf[i]);
+    return failed ? NULL : (byte *)(pc + ldstex_end);
+}
+
+byte *
+decode_common(dcontext_t *dcontext, byte *pc, byte *orig_pc, instr_t *instr)
+{
+    if (DYNAMO_OPTION(build_ldstex)) {
```

Comment:
I'll have to think about that. It would be a different approach. I also have another different approach in mind. So three possible work-arounds to think about. No definite decision yet.